presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java

```diff
@@ -530,19 +530,61 @@ public void setOutput(Optional<Output> output)
         this.output.set(output);
     }
 
-    private void addSerializedCommitOutputToOutput(ConnectorCommitHandle commitHandle)
+    private void addSerializedCommitOutputToOutput(List<?> commitHandles)
     {
         if (!output.get().isPresent()) {
             return;
         }
-
         Output outputInfo = output.get().get();
         SchemaTableName table = new SchemaTableName(outputInfo.getSchema(), outputInfo.getTable());
-        output.set(Optional.of(new Output(
-                outputInfo.getConnectorId(),
-                outputInfo.getSchema(),
-                outputInfo.getTable(),
-                commitHandle.getSerializedCommitOutput(table))));
+        for (int i = 0; i < commitHandles.size(); i++) {
+            if (!(commitHandles.get(i) instanceof ConnectorCommitHandle)) {
+                continue;
+            }
+            ConnectorCommitHandle commitHandle = (ConnectorCommitHandle) commitHandles.get(i);
+            if (commitHandle.hasCommitOutputFor(table)) {
+                output.set(Optional.of(new Output(
+                        outputInfo.getConnectorId(),
+                        outputInfo.getSchema(),
+                        outputInfo.getTable(),
+                        commitHandle.getSerializedCommitOutputForWrite(table))));
+                return;
+            }
+        }
+    }
+
+    private void addSerializedCommitOutputToInputs(List<?> commitHandles)
+    {
+        ImmutableSet.Builder<Input> builder = ImmutableSet.builder();
+
+        for (Input input : inputs.get()) {
+            builder.add(attachSerializedCommitOutput(input, commitHandles));
+        }
+
+        inputs.set(builder.build());
+    }
+
+    private Input attachSerializedCommitOutput(Input input, List<?> commitHandles)
+    {
+        SchemaTableName table = new SchemaTableName(input.getSchema(), input.getTable());
+        for (int i = 0; i < commitHandles.size(); i++) {
+            if (!(commitHandles.get(i) instanceof ConnectorCommitHandle)) {
+                continue;
+            }
+
+            ConnectorCommitHandle commitHandle = (ConnectorCommitHandle) commitHandles.get(i);
+            if (commitHandle.hasCommitOutputFor(table)) {
```

Comment:
Do we need this method? We can just directly create the Input or Output object without checking